src/http/endpoint_key.rs

```diff
@@ -0,0 +1,196 @@
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+use std::ops::Deref;
+
+#[derive(Debug)]
+pub struct EndpointKey(reqwest::Request, Vec<String>);
+
+impl EndpointKey {
+  pub fn new(req: reqwest::Request, headers: Vec<String>) -> Self {
+    EndpointKey(req, headers)
+  }
+  pub fn request(&self) -> reqwest::Request {
+    self.clone().0
+  }
+  pub fn key_headers(&self) -> &Vec<String> {
+    &self.1
+  }
+}
+impl Deref for EndpointKey {
+  type Target = reqwest::Request;
+
+  fn deref(&self) -> &Self::Target {
+    &self.0
+  }
+}
+impl Hash for EndpointKey {
+  fn hash<H: Hasher>(&self, state: &mut H) {
+    self.url().hash(state);
+    self.method().hash(state);
+    for (name, value) in self
+      .headers()
+      .iter()
+      .filter(|(name, _)| self.1.contains(&name.as_str().to_string()))
```

Comment:
Iterate over the allowed headers and check if they exist in the request. If the allowed header count is > 1 then convert it to a set.